src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp

```diff
@@ -422,17 +422,58 @@ void ShenandoahBarrierSet::arraycopy_barrier(T* src, T* dst, size_t count) {
       // Note that we can't do the arraycopy marking using the 'src' array when
       // SATB mode is enabled (so we can't do this as part of the iteration for
       // evacuation or update references).
-      arraycopy_marking(src, dst, count);
+      arraycopy_marking(src, dst, count, true);
     }
   }
 }
 
 template <class T>
-void ShenandoahBarrierSet::arraycopy_marking(T* src, T* dst, size_t count) {
+void ShenandoahBarrierSet::arraycopy_marking(T* src, T* dst, size_t count, bool is_old) {
   assert(_heap->is_concurrent_mark_in_progress(), "only during marking");
-  T* array = ShenandoahSATBBarrier ? dst : src;
-  if (!_heap->marking_context()->allocated_after_mark_start(reinterpret_cast<HeapWord*>(array))) {
-    arraycopy_work<T, false, false, true>(array, count);
+  /*
+   * Note that an old-gen object is considered live if it is live at the start of OLD marking or if it is promoted
+   * following the start of OLD marking.
+   *
+   * 1. Every object promoted following the start of OLD marking will be above TAMS within its old-gen region
+   * 2. Every object live at the start of OLD marking will be referenced from a "root" or it will be referenced from
+   *    another live OLD-gen object.  With regards to old-gen, roots include stack locations and all of live young-gen.
+   *    All root references to old-gen are identified during a bootstrap young collection.  All references from other
+   *    old-gen objects will be marked during the traversal of all old objects, or will be marked by the SATB barrier.
+   *
+   * During old-gen marking (which is interleaved with young-gen collections), call arraycopy_work() if:
+   *
+   * 1. The overwritten array resides in old-gen and it is below TAMS within its old-gen region
+   * 2. Do not call arraycopy_work for any array residing young-gen because young-gen collection is idle at this time
+   *
+   * During young-gen marking, call arraycopy_work() if:
+   *
+   * 1. The overwritten array resides in young-gen and is below TAMS within its young-gen region
+   * 2. Additionally, if array resides in old-gen, regardless of its relationship to TAMS because this old-gen array
+   *    may hold references to young-gen
+   */
+  if (ShenandoahSATBBarrier) {
+    T* array = dst;
+    if (is_old) {
+      ShenandoahHeapRegion* r = _heap->heap_region_containing(array);
+      assert(_heap->mode()->is_generational(), "Only perform old-marking if in generational mode");
+      if (r->is_old() && (((HeapWord*) array)) < _heap->marking_context()->top_at_mark_start(r)) {
```

Comment:
There is a missing `reinterpret_cast<HeapWord*>(array)` here. Probably pull it into the local, so that every branch can just reuse it.

Also, this check is `!_heap->marking_context()->allocated_after_mark_start(reinterpret_cast<HeapWord*>(array))`?